lib/Dialect/BaseModelica/Transforms/CallCSE.cpp

```diff
@@ -47,29 +71,155 @@ mlir::LogicalResult CallCSEPass::processModelOp(ModelOp modelOp) {
+
+  while (!worklist.empty()) {
+    auto *current = worklist.back();
+    worklist.pop_back();
+    toClone.push_back(current);
+    for (auto operand : current->getOperands()) {
+      if (auto *defOp = operand.getDefiningOp()) {
+        worklist.push_back(defOp);
+      }
+    }
+  }
+
+  mlir::Operation *root = nullptr;
+  for (auto *opToClone : llvm::reverse(toClone)) {
+    root = rewriter.clone(*opToClone, mapping);
+  }
+  return root;
+  // for (auto operand : op->getOperands()) {
+  //   if (auto *defOp = operand.getDefiningOp()) {
+  //     cloneDefUseChain(defOp, mapping, rewriter);
+  //   }
+  // }
+
+  // return rewriter.clone(*op, mapping);
+}
+
+EquationTemplateOp
+CallCSEPass::emitCse(ModelOp modelOp, const int emittedCSEs,
+                     llvm::SmallVector<CallOp> &equivalenceGroup,
+                     mlir::RewriterBase &rewriter) {
+  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
+  auto representative = equivalenceGroup.front();
+  // Emit CSE variable
+  rewriter.setInsertionPointToStart(modelOp.getBody());
+  auto cseVariable = rewriter.create<VariableOp>(
+      rewriter.getUnknownLoc(), "_cse" + std::to_string(emittedCSEs),
+      VariableType::wrap(representative.getResult(0).getType()));
+
+  // Create CSE variable driver equation
+  rewriter.setInsertionPointToEnd(modelOp.getBody());
+  auto equationTemplateOp =
+      rewriter.create<EquationTemplateOp>(rewriter.getUnknownLoc());
+  rewriter.setInsertionPointToStart(equationTemplateOp.createBody(0));
+
+  mlir::IRMapping mapping;
+  auto *clonedRepresentative =
+      cloneDefUseChain(representative, mapping, rewriter);
+
+  const auto cseGetOp =
+      rewriter.create<VariableGetOp>(rewriter.getUnknownLoc(), cseVariable);
+  auto lhsOp = rewriter.create<EquationSideOp>(rewriter.getUnknownLoc(),
+                                               cseGetOp->getResults());
+  auto rhsOp = rewriter.create<EquationSideOp>(
+      rewriter.getUnknownLoc(), clonedRepresentative->getResults());
+  rewriter.create<EquationSidesOp>(rewriter.getUnknownLoc(), lhsOp, rhsOp);
+
+  // Replace calls with get to CSE variable
+  for (auto &callOp : equivalenceGroup) {
+    rewriter.setInsertionPoint(callOp);
+    rewriter.replaceOpWithNewOp<VariableGetOp>(callOp, cseVariable);
+  }
+
+  return equationTemplateOp;
+}
+
+mlir::LogicalResult CallCSEPass::processModelOp(ModelOp modelOp) {
+  mlir::IRRewriter rewriter(modelOp);
+  mlir::SymbolTableCollection symbolTableCollection;
+  symbolTableCollection.getSymbolTable(modelOp);
 
   llvm::SmallVector<CallOp> callOps;
+  collectCallOps(modelOp, callOps);
+
+  llvm::SmallVector<llvm::SmallVector<CallOp>> callEquivalenceGroups;
+  buildCallEquivalenceGroups(callOps, symbolTableCollection,
+                             callEquivalenceGroups);
 
-  for (EquationTemplateOp templateOp : templateOps) {
-    templateOp->walk([&](CallOp callOp) {
-      callOps.push_back(callOp);
-    });
+  int emittedCSEs = 0;
```

Comment:
It should be possible - and would be nice - to add this measurement to the [pass statistics](https://mlir.llvm.org/docs/PassManagement/#pass-statistics).